src/toxcore/state_format/old.rs

```diff
@@ -1055,38 +870,110 @@ impl SectionData {
         // TODO: don't return an empty Vec ?
         s.iter()
             .map(|sd| sd.as_section())
-            .filter(|s| s.is_some())
-            .map(|s| s.expect("IS Some(_)"))
+            .filter(|s| s.clone().is_done())
+            .map(|s| s.to_result().unwrap())
             .collect()
     }
 }
 
 
-from_bytes!(SectionData, do_parse!(
-    data_len: le_u32 >>
-    kind: call!(SectionKind::parse_bytes) >>
-    tag!(SECTION_MAGIC) >>
-    data: take!(data_len) >>
-    (SectionData { kind, data: data.to_vec() })
-));
+impl FromBytes for SectionData {
+    named!(from_bytes<SectionData>, do_parse!(
+        data_len: le_u32 >>
+        kind: call!(SectionKind::from_bytes) >>
+        tag!(SECTION_MAGIC) >>
+        data: take!(data_len) >>
+        (SectionData { kind, data: data.to_vec() })
+    ));
+}
+
+#[cfg(test)]
+impl Arbitrary for SectionKind {
+    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+        let mut range_val: u8 = 7;
+
+        while range_val > 6 && range_val < 0x0a {
+            range_val = g.gen_range(1, 0x0c);
+        }
 
+        match range_val {
+            1 => SectionKind::NospamKeys,
+            2 => SectionKind::DHT,
+            3 => SectionKind::Friends,
+            4 => SectionKind::Name,
+            5 => SectionKind::StatusMsg,
+            6 => SectionKind::Status,
+            0x0a => SectionKind::TcpRelays,
+            0x0b => SectionKind::PathNodes,
+            _ => {
+                debug!("system error: gen_range() error");
+                SectionKind::NospamKeys
+            },
+        }
+    }
+}
 
 #[cfg(test)]
 impl Arbitrary for SectionData {
     fn arbitrary<G: Gen>(g: &mut G) -> Self {
-        let kind: SectionKind = Arbitrary::arbitrary(g);
-        let data = match kind {
-            SectionKind::NospamKeys => NospamKeys::arbitrary(g).to_bytes(),
-            SectionKind::DHT => DhtState::arbitrary(g).to_bytes(),
-            SectionKind::Friends => Friends::arbitrary(g).to_bytes(),
-            SectionKind::Name => Name::arbitrary(g).0,
-            SectionKind::StatusMsg => StatusMsg::arbitrary(g).0,
-            SectionKind::Status => vec![UserStatus::arbitrary(g) as u8],
-            SectionKind::TcpRelays => TcpRelays::arbitrary(g).to_bytes(),
-            SectionKind::PathNodes => PathNodes::arbitrary(g).to_bytes(),
-            SectionKind::EOF => vec![],
-        };
-        SectionData { kind: kind, data: data }
+        let mut range_val: u8 = 7;
```

Comment:
The previous implementation was much cleaner. Why did you rewrite it in a such way?

Reply:
Oh! I am sorry. old.rs is not used.
Now new.rs is included.